### DISCUSSION TITLE
/hyndsight/ljung-box-test/

### DISCUSSION THREAD

**Discussion by robjhyndman** [robjhyndman/robjhyndman.com#157] [Announcements]
Thoughts on the Ljung-Box test 
 https://robjhyndman.com/hyndsight/ljung-box-test/

**Comment by robjhyndman** [comment]
```header
Originally posted by Pedro H. C. Sant'Anna on 2014-01-24 19:51:58
````
An interesting approach of selecting the lags h is to use a data-driven
procedure such as the one propose by Escanciano and Lobato (2009,
Journal of Econometrics - An automatic Portmanteau test for serial
correlation)

Nonetheless, my impression is that this would work well well for raw
data, specially if they are iid as in the example. Nonetheless, when you
have estimated parameters, as is the case when you want to perform some
specification testing, the situation would change, specially if the data
is not iid (say you have some garch errors). This also is true for the
classical LB test\...

In these cases, the rule of thumb of using h=square root of n might
provide a decent approximation, but for small values of h, the LB-text
might perform bad.

Good post and nice discussion.

**Comment by robjhyndman** [comment]
```header
Originally posted by Tao Hong on 2014-01-25 21:28:47
````
&quot;we expect the resid­u­als to fail this test when we have suf­fi­cient
data.&quot;\
I think this is a very important remark for practitioners. In load
forecasting for example, we often use several years of hourly or
sub-hourly data to develop models. Many people get nervous when seeing
failures of the test. Then a false action is often to arbitrarily start
adding more variables to make the model more and more complicated, which
may very likes lead to over-fit and create even more troubles in the
forecasts.

**Comment by robjhyndman** [comment]
```header
Originally posted by Kinzie on 2016-02-22 09:37:15
````
Sir, I have raw data of 3000 data series in 2years period (24months).\
I need to check the chi squared statistic using R.

How could I count ?

**Comment by robjhyndman** [comment]
```header
Originally posted by Gabe on 2016-03-29 18:50:24
````
Sorry to comment on a 2-year-old post, but I was reading up on residual
tests and was hoping you could help clarify something.

I\'ve seen recommendations against using a Ljung-Box test when the model
was fitted using lagged dependent variables and instead suggest using a
Breusch-Godfrey test in that case.\
e.g. see
[http://stats.stackexchange\....](http://stats.stackexchange.com/questions/148004/testing-for-autocorrelation-ljung-box-versus-breusch-godfrey &quot;http://stats.stackexchange.com/questions/148004/testing-for-autocorrelation-ljung-box-versus-breusch-godfrey&quot;)
and references therein.

In your post and in sections 2.6 and 5.4 of your textbook (where you
discuss residual diagnostics), none of the examples used lagged
variables, but there\'s also no explicit statements about strict
exogeneity being required (nor did I see any recommendations against
these tests in the ARIMA chapter). Yet the Ljung &amp; Box paper does seem
to claim their test should be used on the residuals from ARMA modeling.

Thus, I\'m a bit confused and was wondering what your take was on this.
Thanks!

**Reply by robjhyndman** [reply]
```header
Originally posted on 2016-03-29 23:05:57
````
Sorry, I\'ve never looked at the alternative tests.

**Reply by robjhyndman** [reply]
```header
Originally posted by Gabe on 2016-03-29 23:57:18
````
Oh ok, thanks for the reply!

For personal reference, then in your experience you\'ve just used/seen
people use the Ljung-Box test to check the residuals after ARMA
modeling?

**Reply by robjhyndman** [reply]
```header
Originally posted on 2016-03-30 01:07:51
````
Frequently.

**Comment by robjhyndman** [comment]
```header
Originally posted by justin on 2016-08-05 18:50:59
````
Hi there, Great post! Thanks. What are your opinions on when to use
portmanteau versus LM tests both in AR and VAR processes?

**Reply by robjhyndman** [reply]
```header
Originally posted on 2016-08-06 02:55:57
````
They have very different purposes. A portmanteau test is checking
whether residuals are white noise. LM tests are used for model
selection. I never use LM tests myself as I think the AIC is a better
approach.

**Comment by robjhyndman** [comment]
```header
Originally posted by germ on 2017-03-10 10:13:07
````
Now with the update, we can use checkresiduals() for easier application
of the Ljung box test.

Also Prof Rob, I have never seen any examples of Ljung box test for any
ets modelling. Is it my misconception? Or we can do the Ljung box test
for ets modelling.

**Reply by robjhyndman** [reply]
```header
Originally posted on 2017-03-12 21:07:50
````
You should do a test for white residuals (using Ljung-Box or something
else) for every forecasting model.

**Comment by robjhyndman** [comment]
```header
Originally posted by Dinesh Tuladhar on 2017-08-28 03:09:51
````
Is it common to see Ljung Box test (p value) suggesting \'the residuals
of a ARIMA model is not auto correlated\' while the correlogram
(ACF/PACF) of the residuals do show some significant autocorrelations?\
In such cases how reasonable will it be to rely on the Ljung Box test
result and ignore the results from correlogram, or we should consider
the ACF/PACF plots equally important?

**Reply by robjhyndman** [reply]
```header
Originally posted on 2017-08-28 04:54:15
````
The LB test is a test on several autocorrelations. You might have one
just outside the limits, but the rest well inside, so as a group they
are not significant. You can also have many autocorrelations close to
the critical values, but none beyond them, but as a group they are
significant.

**Reply by robjhyndman** [reply]
```header
Originally posted by Dinesh Tuladhar on 2017-08-28 08:46:15
````
Thank you very much Rob, that makes sense.

**Comment by robjhyndman** [comment]
```header
Originally posted by Arina Olo on 2017-10-23 20:21:42
````
Dear Mr Hyndman,

thank you very much for the post! I have an issue which you could
possibly help me to resolve.

If I do VAR. With a command:

VARselect(DATA\[5:58,\], lag.max = 4, type = &quot;none&quot;)

I get:

\$selection\
AIC(n) HQ(n) SC(n) FPE(n)\
4 1 1 2

Then, in order to check all the VAR models (VAR(1) to VAR(4)) I used
three tests for residual autocorrelations:

\# Asymptotic Portmanteau test for serially correlated errors\
\# Portmanteau Test (adjusted) for small samples

serial.test(var1,
[lags.pt](http://lags.pt &quot;lags.pt&quot;) = 16, type
= &quot;PT.adjusted&quot;)#serial correlation\
serial.test(var2,
[lags.pt](http://lags.pt &quot;lags.pt&quot;) = 16, type
= &quot;PT.adjusted&quot;)#no serial correlation on 10%\
serial.test(var3,
[lags.pt](http://lags.pt &quot;lags.pt&quot;) = 16, type
= &quot;PT.adjusted&quot;)#serial correlation\
serial.test(var4,
[lags.pt](http://lags.pt &quot;lags.pt&quot;) = 16, type
= &quot;PT.adjusted&quot;)#serial correlation

\# Breusch-Godfrey LM test for small samples (Edgerton-Shukur F test)\
serial.test(var1,
[lags.bg](http://lags.bg &quot;lags.bg&quot;) = 5, type =
&quot;ES&quot;)# no serial correlation\
serial.test(var2,
[lags.bg](http://lags.bg &quot;lags.bg&quot;) = 5, type =
&quot;ES&quot;)# no serial correlation\
serial.test(var3,
[lags.bg](http://lags.bg &quot;lags.bg&quot;) = 5, type =
&quot;ES&quot;)# no serial correlation\
serial.test(var4,
[lags.bg](http://lags.bg &quot;lags.bg&quot;) = 5, type =
&quot;ES&quot;)# no serial correlation

##Test for Autocorrelations\
#H0=No autocorrelation\
Box.test(resid1\[,1\],lag=3,type=&quot;Ljung-Box&quot;)# No Autocorrelation\
Box.test(resid1\[,2\],lag=3,type=&quot;Ljung-Box&quot;)# Autocorrelation on 5%\
Box.test(resid1\[,3\],lag=3,type=&quot;Ljung-Box&quot;)# No Autocorrelation\
Box.test(resid1\[,4\],lag=3,type=&quot;Ljung-Box&quot;)# No Autocorrelation\
Box.test(resid1\[,5\],lag=3,type=&quot;Ljung-Box&quot;)# Autocorrelation on 5%

Box.test(resid2\[,1\],lag=3,type=&quot;Ljung-Box&quot;)# No Autocorrelation\
Box.test(resid2\[,2\],lag=3,type=&quot;Ljung-Box&quot;)# No Autocorrelation\
Box.test(resid2\[,3\],lag=3,type=&quot;Ljung-Box&quot;)# No Autocorrelation\
Box.test(resid2\[,4\],lag=3,type=&quot;Ljung-Box&quot;)# No Autocorrelation\
Box.test(resid2\[,5\],lag=3,type=&quot;Ljung-Box&quot;)# No Autocorrelation

Box.test(resid3\[,1\],lag=3,type=&quot;Ljung-Box&quot;)# No Autocorrelation\
Box.test(resid3\[,2\],lag=3,type=&quot;Ljung-Box&quot;)# No Autocorrelation\
Box.test(resid3\[,3\],lag=3,type=&quot;Ljung-Box&quot;)# No Autocorrelation\
Box.test(resid3\[,4\],lag=3,type=&quot;Ljung-Box&quot;)# No Autocorrelation\
Box.test(resid3\[,5\],lag=3,type=&quot;Ljung-Box&quot;)# No Autocorrelation

Box.test(resid4\[,1\],lag=3,type=&quot;Ljung-Box&quot;)# No Autocorrelation\
Box.test(resid4\[,2\],lag=3,type=&quot;Ljung-Box&quot;)# No Autocorrelation\
Box.test(resid4\[,3\],lag=3,type=&quot;Ljung-Box&quot;)# No Autocorrelation\
Box.test(resid4\[,4\],lag=3,type=&quot;Ljung-Box&quot;)# No Autocorrelation\
Box.test(resid4\[,5\],lag=3,type=&quot;Ljung-Box&quot;)# No Autocorrelation

The 1st problem: what lag length (h) should I use in the tests? The
results differ from lag to lag.\
Secondly, I am not sure if I should use the constant. Is it true that
one could see it from the ACFs and PACFs, both which h to use and if
there should be a constant in the VAR model?

Many thanks in advance for your help!

**Reply by robjhyndman** [reply]
```header
Originally posted on 2017-10-23 22:47:12
````
Don\'t test the five series separately. You need a joint multivariate
white noise test. See a book on multivariate time series such as
Lutkepohl or Reinsel.

**Comment by robjhyndman** [comment]
```header
Originally posted by ezequiel boehler on 2018-02-22 22:40:46
````
Greetings Prof. Hyndman.

Would you be so kind to explain to me what could mean having a high
correlation and small p-values when running AFC and Ljung on the raw
data, but at the same time having high correlation and small p-values on
the residuals of that same data?

Thank you so much in advance!

**Reply by robjhyndman** [reply]
```header
Originally posted on 2018-02-22 23:05:58
````
The model is not very good.

**Reply by robjhyndman** [reply]
```header
Originally posted by ezequiel boehler on 2018-02-23 03:59:36
````
Great! Was not sure about that. Thank you!

**Comment by robjhyndman** [comment]
```header
Originally posted by Miles McBain on 2018-05-03 05:34:18
````
Hi Rob, I noticed the code shipped with forecast varies a little bit
from this advice and I was wondering what had changed in your thinking?
See:
[https://github.com/robjhynd\...](https://github.com/robjhyndman/forecast/blob/e98f7b5519831d936a1db2c1ac0598f722d86050/R/checkresiduals.R#L124 &quot;https://github.com/robjhyndman/forecast/blob/e98f7b5519831d936a1db2c1ac0598f722d86050/R/checkresiduals.R#L124&quot;)

\`df + 3\` stuck out to me. For an ARIMA the code treats this as K + 3.
I can make some hand waving justifications involving propensity to use
model selection but am largely at a loss to explain it myself.

**Reply by robjhyndman** [reply]
```header
Originally posted on 2018-05-03 09:31:00
````
The test uses a chisquare dist with lag-df degrees of freedom, so lags
has to be at least df+1. I think I probably chose to make it a minimum
of df+3 simply because a portmanteau test of multiple lags doesn\'t mean
much unless it has at least a few lags involved. But I can\'t recall why
we don\'t have the T/5 limit. Most likely I forgot about this post when
I wrote the checkresiduals() function! I\'ll fix it.

**Comment by robjhyndman** [comment]
```header
Originally posted by Lita Lertlumprasert on 2018-07-11 21:33:36
````
Hi there,

From the 2nd paragraph where it says &quot;K is the number of parameters
estimated in the model,&quot; does anyone know what my K would be if my
model is an ensemble model of 5 ML models and 4 time-series models
(ARIMA, STL, Snaive, ETS, KNN, ranger, glmnet, SVMradial, and XGBtree).
The ensemble method picks the &quot;best forecast&quot; based on historical
errors of each model at specific horizons. Would K be the number of
hyperparameters from all of the models used? Or would it be the number
of models used for the ensemble model?

Any help would be much appreciated.

Thank you,\
Lita

**Reply by robjhyndman** [reply]
```header
Originally posted on 2018-07-11 22:04:13
````
Great question, and I\'m not sure of the answer. A reasonable guess
might be the average number of parameters across the nine models. It is
also not clear what K is for the ML models. It should be the number of
df in the algorithm, but that is not always clear either.

**Comment by robjhyndman** [comment]
```header
Originally posted by davidreilly7 on 2019-02-25 15:02:08
````
Hi Rob,

Running ets with log AirPassngers, followed by checkresiduals gives very
low Ljung-Box p-values. Perplexed.

Thanks!

library(forecast)\
ap=ts(AirPassengers,start=c(1949,1),frequency=12)\
log.AP = log(ap)

[ets.fit](http://ets.fit &quot;ets.fit&quot;) \&lt;-
ets(ap,lambda=0)\
checkresiduals([ets.fit](http://ets.fit &quot;ets.fit&quot;))

[ets.fit](http://ets.fit &quot;ets.fit&quot;) \&lt;-
ets(log.AP)\
checkresiduals([ets.fit](http://ets.fit &quot;ets.fit&quot;))

\&gt; [ets.fit](http://ets.fit &quot;ets.fit&quot;) \&lt;-
ets(ap,lambda=0)\
\&gt;
checkresiduals([ets.fit](http://ets.fit &quot;ets.fit&quot;))

Ljung-Box test

data: Residuals from ETS(A,A,A)\
Q\* = 53.567, df = 8, p-value = 8.376e-09

Model df: 16. Total lags used: 24

\&gt; [ets.fit](http://ets.fit &quot;ets.fit&quot;) \&lt;-
ets(log.AP)\
\&gt;
checkresiduals([ets.fit](http://ets.fit &quot;ets.fit&quot;))

Ljung-Box test

data: Residuals from ETS(M,A,M)\
Q\* = 45.416, df = 8, p-value = 3.069e-07

Model df: 16. Total lags used: 24

**Reply by robjhyndman** [reply]
```header
Originally posted on 2019-02-25 20:40:23
````
The ETS model space is relatively small and it is common to find that
the LB test fails on residuals even though the forecasts might be good.

**Reply by robjhyndman** [reply]
```header
Originally posted by davidreilly7 on 2019-02-25 20:45:07
````
Ok thank you for the explanation!

Can one generalize that LB is more suited for use with ARIMA models?

**Reply by robjhyndman** [reply]
```header
Originally posted on 2019-02-25 20:56:04
````
No. It is a test for correlated residuals. But failing the test does not
necessarily mean that (a) the model produces poor forecasts; or (b) that
the prediction intervals are in accurate. It suggests that there is a
little more information in the data than is captured in the model. But
it might not matter much.

**Reply by robjhyndman** [reply]
```header
Originally posted by davidreilly7 on 2019-02-25 21:00:19
````
Thank you very much! I recall a similar situation in a regression model
where there was significant lack of fit, but R-Square was very high and
the model predicted very well.

**Comment by robjhyndman** [comment]
```header
Originally posted by Konsta on 2019-03-12 12:53:15
````
The link to Burns\' online paper is dead. Is it still available
somewhere?

**Reply by robjhyndman** [reply]
```header
Originally posted on 2019-03-12 21:29:44
````
I\'ve fixed the link to point to its new location.

**Comment by robjhyndman** [comment]
```header
Originally posted by Duaa Taleb on 2019-04-08 05:15:57
````
Hi prof,\
some times when i apply ljung-box test in R, the thierd shape show p.v
after some lags &quot;it start from lag5&quot; what is the problem?

**Comment by Diogo-loureiro** [comment]
Hi Professor. I have one question. Does it make sense to apply this test in the time series before trying to create a forecasting model? I've seen some practioners suggesting to apply this test in the series before start working in the model. If the test shows that the time series is white noise, anything more complex than a mean would be a waste. Does it make sense? I couldn't find any books backing up this practice.

**Reply by robjhyndman** [reply]
Yes, you can do that.
